rfcs/2020-11-06-4909-remap-template-fields.md

```diff
@@ -0,0 +1,191 @@
+fields are only removed from the event at the end. This differs from the `del`
+function in that `del` will remove the field immediately and it would not be
+available for use after that point. The order in which the template fields are
+calculated would have an impact on the final result.
+
+
+## Rationale
+
+The benefits of using Remap in the template fields are:
+
+ - One familiar syntax and function reference for Vector.
+ - Access to all of remap's functions for templating.
+ - Less code to manage (once the old template fields are deprecated).
+
+
+## Drawbacks
+
+There could be an additional maintenance burden. The Remap syntax is more
+complex which does mean there is more of a learning curve for the user and more
+likelyhood that they make mistakes.
+
+
+## Alternatives
+
+### Do nothing
+
+We do already have the existing template syntax. Perhaps we can stick with this.
+
+The advantage of using Remap for these fields are that it allows more 
+flexibility in defining how the event is used. However, given that remap can be 
+used as a transform, should the user really need this they could put a Remap 
+transform in the process to process these fields so they can be easily used in 
+the template for the next phase.
+
+### Use distinct syntax to distinguish between template and Remap syntax.
+
+We could specify that double parentheses (`{{..}}`) are used for template
+syntax (the current technique) and triple parentheses (`{{{...}}}`) are used 
+for Remap syntax.
+
+This would provide a clear and unambiguous way to distinguish between the
+syntaxes rather than rely on a set of heuristics.
+
+### Returning fields at load time
+
+Instead of requiring the user to mark the fields that are removed, Remap could
+keep track of the fields used whilst running the script.
+
+There are several options:
+
+- *Keep track of all fields used in the script.*
+This script, `if .foo { .bar } else { .baz }`, would result in all three
+fields being returned - `.foo`, `.bar` and `.baz`, and subsequently removed
+from the message sent to Loki.
+
+- *Keep track of the fields used in the run path.*
+The script, `if .foo { .bar } else { .baz }`,
+would result in `.foo` and either `.bar` or `.baz` being returned.
+
+If necessary, Remap could distinguish between fields that are used in the
+condition and those used it the result, so only `.bar` or `.baz` could be
+returned.
+
+There are likely to be a number of edge cases that would need to be thought
+through if we took this approach. For example, if a field is used in a function
+but it's value is only used to influence the result - should it be included?
+
+```
+replace(.field1, .field2, .field3)
+```
+
+The result is the value of `field1`, but with any occurrence of the value of 
+`field2` being replaced by `field3`. Which fields would be correct to include
+in this list?
+
+
+## Outstanding Questions
+
+- How important is it for the Loki sink to remove fields used in the template
```

Comment:
I'd propose deprecating `remove_label_fields`; allowing its use only with non-remap templates.

I think, going forward, a better way to handle options like `remove_label_fields` is to let users use `remap` in the component itself. That is, they'd do something like:

```toml
[sinks.loki]
  type = "loki"
  inputs = ["json"]
  endpoint = "http://localhost:3100"
  remove_label_fields = true
  labels.key = "{{{ .foo }}}-{{{ .buzz }}}"
  remap = """del(.foo, .buzz)"""
```

Reply:
I agree with this. This is a clearly superior solution 👍

I'd probably vote to have two config fields available in _all_ components:

* `remap.before`
* `remap.after`

Both take a Remap script, and run either before or after the other logic of the component finished.

Reply:
> I'd probably vote to have two config fields available in all components:

Curious what @lukesteensen thinks about this. We've had similar discussions for control flow:

```toml
[transforms.sampler]
  type = "sampler"
  
  [[if]]
    condition = ".status == 200"
    # regular component options here.

  [[if]]
    condition = ".status == 500"
    # regular component options here.
```